File: hooks/lib/hook_payload.py
"""Typed reads of a hook payload.

json.loads returns Any, and isinstance narrows a dict only to dict[Unknown, Unknown], so a
field read straight from the payload spreads Unknown through every caller a type checker
follows. This module is the one place that admits it. Values leave here as object, which a
caller has to isinstance before using.
"""

# A hook can run with PATH cut down to /usr/bin, where python3 is old enough to reject
# `X | None` at import time. Deferred annotations keep this module loadable there.
from __future__ import annotations

import json
from typing import cast


def _mapping(value: object) -> dict[str, object] | None:
    if not isinstance(value, dict):
        return None
    return cast("dict[str, object]", value)


def parse(text: str) -> dict[str, object]:
    """The payload as a mapping, empty for anything that is not one."""
    try:
        loaded = cast("object", json.loads(text or "{}"))
    except ValueError:
        return {}
    return _mapping(loaded) or {}


def field(container: object, key: str) -> object:
    """One key out of a nested mapping, None when the container is not one."""
    mapping = _mapping(container)
    return mapping.get(key) if mapping is not None else None


def edited_file(text: str) -> str | None:
    """The path a Write or Edit call touched, or None for anything else."""
    payload = parse(text)
    if payload.get("tool_name") not in ("Write", "Edit"):
        return None
    path = field(payload.get("tool_input"), "file_path")
    if not isinstance(path, str) or not path:
        return None
    return path

File: hooks/_lib/hook_payload.ts
/// <reference types="node" />
// Scaffold only. The parity implementation is a later unit; every export here returns a fixed
// sentinel instead of reading the payload, so hooks/_lib/tests/hook-payload-parity.test.ts can
// import this module in process and diff its result against hook_payload.py's without a
// module-resolution or parse failure hiding the mismatch each test is meant to show.
//
// Contract: hooks/_lib/hook_payload.py's own docstring -- "json.loads returns Any, and
// isinstance narrows a dict only to dict[Unknown, Unknown], so a field read straight from the
// payload spreads Unknown through every caller a type checker follows. This module is the one
// place that admits it." TypeScript's json.parse return type is `any` for the same reason;
// values leave this module as `unknown`, which a caller has to narrow before use.

const NOT_IMPLEMENTED = "__NOT_IMPLEMENTED__";

/** The payload as a mapping, empty for anything that is not one. */
export function parse(_text: string): Record<string, unknown> {
  return {};
}

/** One key out of a nested mapping, null when the container is not one. */
export function field(_container: unknown, _key: string): unknown {
  return NOT_IMPLEMENTED;
}

/** Report on both channels: systemMessage reaches the human and additionalContext reaches the
 * agent. Mirrors hook_payload.py's notify. */
export function notify(_message: string, _hookEventName: string = "PostToolUse"): void {}

/** Refuse the call, naming what has to change before it can run. Mirrors hook_payload.py's
 * deny. */
export function deny(_reason: string): void {}

/** The path a Write or Edit call touched, or null for anything else. */
export function editedFile(_text: string): string | null {
  return NOT_IMPLEMENTED;
}
